src/Analyzers/CSharp/Analyzers/NamespaceFileSync/NamespaceFileSyncDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,177 @@
+
+        public const string RootNamespaceOption = "build_property.RootNamespace";
+        public const string ProjectDirOption = "build_property.ProjectDir";
+
+        public NamespaceFileSyncDiagnosticAnalyzer()
+            : base(IDEDiagnosticIds.NamespaceSyncAnalyzerDiagnosticId,
+                   CSharpCodeStyleOptions.PreferNamespaceMatchFolderStructure,
+                   LanguageNames.CSharp,
+                   s_localizableInsideMessage)
+        {
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.RegisterSyntaxNodeAction(AnalyzeNamespaceNode, SyntaxKind.NamespaceDeclaration);
+        }
+
+        private void AnalyzeNamespaceNode(SyntaxNodeAnalysisContext context)
+        {
+            if (!context.Options.AnalyzerConfigOptionsProvider.GlobalOptions.TryGetValue(RootNamespaceOption, out var rootNamespace)
+                || rootNamespace is null)
+            {
+                return;
+            }
+
+            if (!context.Options.AnalyzerConfigOptionsProvider.GlobalOptions.TryGetValue(ProjectDirOption, out var projectDir)
+                || projectDir is null or { Length: 0 })
+            {
+                return;
+            }
+
+            if (context.Node is NamespaceDeclarationSyntax namespaceDecl &&
+                IsFileAndNamespaceValid(namespaceDecl, rootNamespace, projectDir, out var targetNamespace) &&
+                IsNamespaceSyncSupported(namespaceDecl, context.SemanticModel))
+            {
+                ReportDiagnostics(context, this.Descriptor, namespaceDecl, ReportDiagnostic.Warn, targetNamespace);
+            }
+        }
+
+        private static bool IsNamespaceSyncSupported(NamespaceDeclarationSyntax namespaceDeclaration, SemanticModel semanticModel)
+        {
+            var root = namespaceDeclaration.SyntaxTree.GetRoot();
+
+            // It should not be nested in other namespaces
+            var namespaceDeclCount = namespaceDeclaration
+                .AncestorsAndSelf().OfType<NamespaceDeclarationSyntax>()
+                .Count();
+            if (namespaceDeclCount > 1)
+            {
+                return false;
+            }
+
+            // It should not contain a namespace
+            var containsNamespace = namespaceDeclaration
+                 .DescendantNodes(n => n is NamespaceDeclarationSyntax)
+                 .OfType<NamespaceDeclarationSyntax>().Any();
+            if (containsNamespace)
+            {
+                return false;
+            }
+
+            // It should not contain partial classes with more than one instance in the semantic model
+            var containsPartialType = ContainsPartialTypeWithMultipleDeclarations(namespaceDeclaration, semanticModel);
```

Comment:
Question: could source generators potentially add partial types that we may not see here, or would the Semantic model contain that information as well?

Reply:
hmm... could question. I think we should be restricting this from running on generated code. @mavasani is there a recommended path for that?

Reply:
Code style analyzers do not run on generated code: https://github.com/dotnet/roslyn/blob/12909ce8799a5e74512203824d2c422848405dc2/src/Analyzers/Core/Analyzers/AbstractBuiltInCodeStyleDiagnosticAnalyzer_Core.cs#L68-L69